app/code/Magento/Backend/Block/Widget/Grid/Column/Renderer/Options.php

```diff
@@ -43,7 +43,7 @@ public function render(\Magento\Framework\DataObject $row)
                 $output[$option['value']] = $option['label'];
             }
 
-            $value = $row->getData($this->getColumn()->getIndex());
+            $value = (int)$row->getData($this->getColumn()->getIndex());
```

Comment:
I think that converting value to int could broke functionality for sources where values is not int. Values could be `some_string_value`.
I think problem is more difficult. I checked in debugger and see that `_loadAttributes` use `union` select for selecting all attributes in one select. Use of `union` imposes a limitation for columns types. In this case int values convert to decimal on reordering query. On page load values is correct because it's converting to varchar

Reply:
@Den4ik, thanks for your answer! I rushed to the decision, you are right. I think we should change a select to database, not f result of the query. Maybe we should to add some attribute from category_product_entity_varchar table. Maybe it helps to get string values in response. For example we can add 'name' attribute separately to get new SELECT in UNION ALL query.

Reply:
@SilinMykola In my opinion adding `name` to select is ok as fast solution